File: solutions/kotlin/pig-latin/src/main/kotlin/PigLatin.kt
object PigLatin {
    private val vowels = listOf("a", "e", "i", "o", "u", "xr", "yt")
    private val consonants = ('a'..'z')
            .map { it.toString() }
            .filterNot { it in vowels }
            .plus(listOf("ch", "qu", "th", "rh", "thr", "sch"))

    fun translate(input: String) = input
            .split(" ")
            .joinToString(" ") { it.toPig() }

    private fun String.toPig() = when {
        take(1) in vowels || take(2) in vowels -> this + "ay"
        drop(1).take(2) == "qu" -> drop(3) + take(3) + "ay"
        take(3) in consonants -> drop(3) + subSequence(0, 3) + "ay"
        take(2) in consonants -> drop(2) + subSequence(0, 2) + "ay"
        take(1) in consonants -> drop(1) + take(1) + "ay"
        else -> this
    }
}

File: solutions/scala/pig-latin/src/main/scala/PigLatin.scala
object PigLatin {
    def pigIt(t: String): Int = t.toSeq match {
        case Seq('a' | 'e' | 'i' | 'o' | 'u', _*) => 0
        case Seq('q', 'u', _*) => 2
        case Seq('c' | 'r', 'h', _*) => 2
        case Seq('t', 'h', 'r', _*) => 3
        case Seq('t', 'h', _*) => 2
        case Seq('x', 'r', _*) => 0
        case Seq('y', 't', _*) => 0
        case Seq('s', 'c', 'h', _*) => 3
        case Seq('s', 'q', 'u', _*) => 3
        case _ => 1
    }

    def translate(txt: String): String = txt.split(" ").map { t =>
        ((n: Int) => t.drop(n) + t.take(n) + "ay")(pigIt(t))}.mkString(" ")
}